src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestChannelInternal.java

```diff
@@ -353,7 +355,16 @@ public InsertValidationResponse insertRows(
       throw new SFException(ErrorCode.CLOSED_CHANNEL, getFullyQualifiedName());
     }
 
-    InsertValidationResponse response = this.rowBuffer.insertRows(rows, offsetToken);
+    // We create a shallow copy to protect against concurrent addition/removal of columns, which can
+    // lead to double counting of null values, for example. Individual mutable values may still be
+    // concurrently modified (e.g. byte[]). Before validation and EP calculation, we must make sure
+    // that defensive copies of all mutable objects are created.
+    final List<Map<String, Object>> rowsCopy = new LinkedList<>();
+    for (Map<String, Object> row : rows) {
+      rowsCopy.add(new HashMap<>(row));
```

Comment:
Nit, can reduce to `rows.forEach(r -> rowsCopy.add(new HashMap<>(r)))`;

Reply:
Will fix in a separate PR, I want to merge this one quickly, so it makes it to the next release.

Reply:
SGTM!